### PR TITLE
[WIP] Add precise scroll event handling

### PR DESCRIPTION
**Pull request by szbergeron**

### CODE REVIEW

File: src/SDL.c [hunk 1]
```diff
@@ -105,6 +105,7 @@ SDL_SetMainReady(void)
 int
 SDL_InitSubSystem(Uint32 flags)
 {
+    printf("our init called\n");
```
Comment (szbergeron) [inline src/SDL.c]: remove prints

File: src/SDL.c [hunk 2]
```diff
@@ -252,6 +253,7 @@ SDL_InitSubSystem(Uint32 flags)
 int
 SDL_Init(Uint32 flags)
 {
+    printf("our init called\n");
```
Comment (szbergeron) [inline src/SDL.c]: remove prints

File: src/events/SDL_mouse.c [hunk 3]
```diff
@@ -188,7 +188,7 @@ SDL_GetMouseFocus(void)
     return mouse->focus;
 }
 
-#if 0
+#if 0 // TODO: document what this does
```
Comment (szbergeron) [inline src/events/SDL_mouse.c:191]: whoops, that's unhelpful. TODO: actually figure out what it does :P

File: src/video/wayland/SDL_waylandevents.c [hunk 4]
```diff
@@ -355,6 +355,7 @@ pointer_handle_axis_common(struct SDL_WaylandInput *input,
     enum wl_pointer_axis a = axis;
     float x, y;
 
+    printf("our axis common getting called\n");
```
Comment (szbergeron) [inline src/video/wayland/SDL_waylandevents.c]: eventually remove prints here

File: src/SDL.c (hunk 1 above)
Comment (szbergeron) [inline src/SDL.c]: prints

File: src/SDL.c (hunk 2 above)
Comment (szbergeron) [inline src/SDL.c]: prints

File: src/video/wayland/SDL_waylandevents.c [hunk 5]
```diff
@@ -883,7 +962,7 @@ Wayland_display_add_input(SDL_VideoData *d, uint32_t id)
         return;
 
     input->display = d;
-    input->seat = wl_registry_bind(d->registry, id, &wl_seat_interface, 1);
+    input->seat = wl_registry_bind(d->registry, id, &wl_seat_interface, 5);
```
Comment (szbergeron) [inline src/video/wayland/SDL_waylandevents.c:964]: @bryphe this breaks backwards compat with some really old wayland compositor versions (not sure if present on some LTS distros) but they're old enough that they're likely not usable over their X based counterparts. Should switching logic be put in here before any merge happens?

Comment (bryphe) [reply src/video/wayland/SDL_waylandevents.c:964]: @szbergeron - I'm unfortunately not an expert on this!

What would the experience be if you had a really old wayland compositor? A crash? And how old are we talking / how likely is it to be hit?

I'd like to make sure it's diagnosable, at least on my end, in the case that we do hit this - perhaps detection + an SDL_Log?

File: src/video/wayland/SDL_waylandevents.c [hunk 6]
```diff
@@ -633,9 +707,15 @@ seat_handle_capabilities(void *data, struct wl_seat *seat,
     }
 }
 
+static void
+seat_handle_name(void *data, struct wl_seat *seat, char *name)
+{
+    printf("seat has name %s\n", name);
```
Comment (bryphe) [inline src/video/wayland/SDL_waylandevents.c]: We might want to move this to go through the `SDL_Log` infra

File: include/SDL_mouse.h [hunk 7]
```diff
@@ -69,6 +69,29 @@ typedef enum
     SDL_MOUSEWHEEL_FLIPPED    /**< The scroll direction is flipped / natural */
 } SDL_MouseWheelDirection;
 
+/**
+ * source_type indicates what kind of device dispatched this wheel event.
```
Comment (bryphe) [inline include/SDL_mouse.h:73]: Thanks for the details!

File: src/video/wayland/SDL_waylandevents.c (hunk 5 above)
Comment (szbergeron) [reply src/video/wayland/SDL_waylandevents.c:964]: @bryphe Hitting this makes the compositor kill the app and print a message saying that wl_seat versioning is incompatible to the console (reproducible by putting, say, a "6" in that position on a newer compositor)

The interface was added to the spec in late 2015 of this versioning (in commit https://github.com/wayland-project/wayland/commit/c5356e9016aa814a873a765bb2cbe57e804e5ea7) so any server newer than mid-2016 or so should at least support requesting that version, if not fully support it. Until middle of this year KDE didn't send out axis_stop events but would still provide for seats v5 requests. The callbacks in question just wouldn't get called :woman_shrugging: 

Not sure if 3.5-ish years of buffer is enough to assume this is available, I'd give some thought to the fact that wayland only really became usable in desktop settings within the past year or so, so unless Oni is going to be run on the built in screen of a car from 2014 we're pretty unlikely to run into this

File: src/video/wayland/SDL_waylandevents.c (hunk 6 above)
Comment (szbergeron) [reply src/video/wayland/SDL_waylandevents.c]: This was mostly just for proto and can probably be stripped out

File: src/video/wayland/SDL_waylandevents.c (hunk 5 above)
Comment (bryphe) [reply src/video/wayland/SDL_waylandevents.c:964]: Thanks for the details! If we get a clear error message, I'm OK with it.

File: src/video/wayland/SDL_waylandevents.c [hunk 8]
```diff
@@ -378,8 +379,80 @@ pointer_handle_axis(void *data, struct wl_pointer *pointer,
...
+static void
+pointer_handle_axis_discrete(void *data, struct wl_pointer *pointer,
+                             uint32_t axis, uint32_t discrete)
+{
+    struct SDL_WaylandInput *input = data;
+    SDL_WindowData *window = input->pointer_focus;
+    Uint64 x = 0;
+    Uint64 y = 0;
+
+    printf("descrite with %d\n", discrete);
```
Comment (bryphe) [inline src/video/wayland/SDL_waylandevents.c]: Looks like there are still some `printfs` scattered here - we should remove before merging.

File: src/video/wayland/SDL_waylandevents.c [hunk 9]
```diff
@@ -355,6 +355,7 @@ pointer_handle_axis_common(struct SDL_WaylandInput *input,
     enum wl_pointer_axis a = axis;
     float x, y;
 
+    printf("axis common, value %d, axis %d\n", value, axis);
```
Comment (bryphe) [inline src/video/wayland/SDL_waylandevents.c]: Looks like some debug logging is still around here - we should remove prior to merging

**Comment by bryphe** [comment]
Thank you @szbergeron ! Looks great 👍

**Pull request merged**